### PR TITLE
[tx] Added lora embed

### PR DESCRIPTION
**Pull request by Guido1Alessandro1Trevisan**
First attempt at adding LoRA support for the embeddings layer.
I’ve verified the forward backward passes and optimizer step everything runs and the loss decreases as expected.

I’m wondering what better tests are there to confirm that LoRA is functioning correctly. Subtle token routing bugs could really ruin someone’s day.

### CODE REVIEW

File: skyrl-tx/tx/layers/lora.py [hunk 1]
```diff
@@ -84,6 +84,74 @@ def apply_lora(
...
+            sharding_A=jax.sharding.PartitionSpec(None, sharding[0], None),
+            sharding_B=jax.sharding.PartitionSpec(None, None, sharding[1]),
+            dtype=param_dtype,
+            rngs=rngs,
+        )
+
+    def __call__(self, x: jax.Array, adapter_indices: jax.Array | None = None) -> jax.Array:
+        base_out = super().__call__(x)
+
+        if self.max_lora_adapters == 0 or adapter_indices is None:
```
Comment (pcmoritz) [inline skyrl-tx/tx/layers/lora.py]: I think here we should re-use self.apply_lora (we might need to parametrize or modify that slightly)

Comment (Guido1Alessandro1Trevisan) [reply skyrl-tx/tx/layers/lora.py]: Definitely. I was hesitant on the first attempt since there are quite a few small differences here and there between the embedding lora and the linear lora. a better apply_lora is something to brood on

**Comment by Guido1Alessandro1Trevisan** [comment]
Bumping this for merge. parametrized apply_lora,  fixed the test and an issue with the lora expert layer.

**Review by pcmoritz — APPROVED** [review]
I simplified the code some more now and added a test, it should be ready to merge! Thanks a lot for the contribution :)
